src/link-preview-expander.ts

```diff
@@ -1,16 +1,75 @@
 import * as sourcegraph from 'sourcegraph'
+import { concat, of } from 'rxjs'
+import { fromFetch } from 'rxjs/fetch'
+import { catchError, map, switchMap } from 'rxjs/operators'
 
-export function activate(ctx: sourcegraph.ExtensionContext): void {
-    ctx.subscriptions.add(
+export function activate(context: sourcegraph.ExtensionContext): void {
+    context.subscriptions.add(
         sourcegraph.languages.registerHoverProvider(['*'], {
-            provideHover: () => ({
-                contents: {
-                    value: 'Hello world from link-preview-expander! 🎉🎉🎉',
-                    kind: sourcegraph.MarkupKind.Markdown
+            provideHover: (document, position) => {
+                const range = document.getWordRangeAtPosition(position)
+                if (!range) {
+                    return null
                 }
-            }),
+                const maybeURL = getWord(document, range)
+                const isURL = checkIsURL(maybeURL)
+
+                if (!isURL) {
+                    return null
+                }
+
+                const createResult: (debugString: string) => sourcegraph.Badged<sourcegraph.Hover> = (
+                    debugString: string
+                ) => ({
+                    contents: {
+                        value: `<p>${debugString}</p>
+                        <a href="${maybeURL}" target="_blank" rel="noopener noreferrer">Navigate to link</a>
+                        <img width="64" src="https://miro.medium.com/max/816/1*mn6bOs7s6Qbao15PMNRyOA.png" />`,
+                        kind: sourcegraph.MarkupKind.Markdown,
+                    },
+                })
+
+                return concat(
+                    of(createResult('loading')),
+                    fromFetch(maybeURL).pipe(
+                        switchMap(response => {
+                            if (response.ok) {
+                                return response.json()
+                            }
+                            return of(createResult('error'))
+                        }),
+                        map(() => createResult('loaded')),
+                        catchError(() => of(createResult('error')))
```

Comment:
You can also just let the Observable error, that way the error is displayed in a red alert.

Reply:
I get `Provider errored: TypeError: Failed to fetch` in the console when I don't catch errors, and no alert is shown

Reply:
Ah, that may be an intentional UX decision we made at some point to not have badly behaving extensions ruin the UI. I think it was just copied from VS Code's behavior. I would still abide to that and if we wanted to change that we can globally.